assert/assertions.go

```diff
@@ -431,22 +454,40 @@ func samePointers(first, second interface{}) bool {
 // to a type conversion in the Go grammar.
 func formatUnequalValues(expected, actual interface{}) (e string, a string) {
 	if reflect.TypeOf(expected) != reflect.TypeOf(actual) {
-		return fmt.Sprintf("%T(%s)", expected, truncatingFormat(expected)),
-			fmt.Sprintf("%T(%s)", actual, truncatingFormat(actual))
+		return fmt.Sprintf("%T(%s)", expected, truncatingFormat(expected, false)),
+			fmt.Sprintf("%T(%s)", actual, truncatingFormat(actual, false))
 	}
 	switch expected.(type) {
 	case time.Duration:
 		return fmt.Sprintf("%v", expected), fmt.Sprintf("%v", actual)
 	}
-	return truncatingFormat(expected), truncatingFormat(actual)
+
+	return truncatingFormat(expected, true), truncatingFormat(actual, true)
 }
 
 // truncatingFormat formats the data and truncates it if it's too long.
 //
 // This helps keep formatted error messages lines from exceeding the
 // bufio.MaxScanTokenSize max line length that the go testing framework imposes.
-func truncatingFormat(data interface{}) string {
-	value := fmt.Sprintf("%#v", data)
+//
+// The `printValueOfBasicTypes` flag controls whether or not to print the underlying
+// value of pointers of basic types in addition to the pointer address.
+func truncatingFormat(data interface{}, printValueOfBasicTypes bool) string {
```

Comment:
Would it not be neater if we just moved the logic for checking whether to print the value into this function rather than passing it in as an argument at every call?

Reply:
The logic for checking whether to print the value is already inside the `truncatingFormat` function (when it loops through the `basicTypes` list and checks if the type of `data` matches any of the types in the `basicTypes` list).

The use of the `printValueOfBasicTypes` is an override, where if it is set to `false`, only the pointer address (and not the underlying value) is returned, regardless of whether it is a basic type or not. The use case of this is in the `formatUnequalValues` function where two pointers of different types are being compared. In this case, we only want the pointer addresses to be printed and not the underlying values (even if the underlying values are basic types). This emphasizes the fact that the pointers are different types (which is why the comparison failed) and that the underlying value of the pointers do not matter for the comparison.

Also updated the comments for the `truncatingFormat` function to better explain the use of the `printValueOfBasicTypes`